app/javascript/lib/validation/expression-matrices-validation.js

```diff
@@ -0,0 +1,265 @@
+/**
+* @fileoverview Functions used for sparse (MTX), features, barcodes and dense file validation
+*/
+
+import {
+  getParsedHeaderLines, parseLine,
+  validateUniqueCellNamesWithinFile, validateMetadataLabelMatches, validateGroupColumnCounts
+} from './shared-validation'
+
+/** Parse a dense matrix file */
+export async function parseDenseMatrixFile(chunker, mimeType, fileOptions) {
+  const { headers, delimiter } = await getParsedHeaderLines(chunker, mimeType)
+  let issues = validateDenseHeader(headers)
+
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
+    const line = parseLine(rawLine, delimiter)
+    issues = issues.concat(validateValuesAreNumeric(line, isLastLine, lineNum, dataObj))
+    issues = issues.concat(validateColumnNumber(line, isLastLine, headers, lineNum, dataObj))
+    issues = issues.concat(validateUniqueCellNamesWithinFile(line, isLastLine, dataObj))
+    issues = issues.concat(validateMetadataLabelMatches(headers, line, isLastLine, dataObj))
+    issues = issues.concat(validateGroupColumnCounts(headers, line, isLastLine, dataObj))
+    // add other line-by-line validations here
+  })
+  return { issues, delimiter, numColumns: headers[0].length }
+}
+
+/** Parse an MTX matrix file */
+export async function parseSparseMatrixFile(chunker, mimeType, fileOptions) {
+  let issues = []
+  const delimiter = '  ' // tab delimited
```

Comment:
Thanks for clarifying this.  Per my other comment mentioning `tabCharacter`, I think we could be even clearer and more robust.

```suggestion
  const delimiter = tabCharacter
```